man/man1/corduroy.1
.\"    This file is part of the FElt finite element analysis package.
.\"    Copyright (C) 1993-2000 Jason I. Gobat and Darren C. Atkinson
.\"
.\"    This program is free software; you can redistribute it and/or modify
.\"    it under the terms of the GNU General Public License as published by
.\"    the Free Software Foundation; either version 2 of the License, or
.\"    (at your option) any later version.
.\"
.\"    This program is distributed in the hope that it will be useful,
.\"    but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"    GNU General Public License for more details.
.\"
.\"    You should have received a copy of the GNU General Public License
.\"    along with this program; if not, write to the Free Software
.\"    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.TH CORDUROY 1fe "2/25/94" "Version 2.1" "Finite Element Package"
.SH NAME
corduroy \- element mesh generator
.SH SYNOPSIS
.B corduroy
[\-help]
[\-debug]
[\-nocpp]
[\-cpp \fIfilename\fR]
[\-D\fIname\fR[=\fIvalue\fR]]
[\-U\fIname\fR]
[\-I\fIdirectory\fR]
[\fIfilename\fR]
.SH DESCRIPTION
\fICorduroy\fR is an element mesh generator with a flexible input file
syntax (see \fIcorduroy\fR(4fe)) that writes a \fIfelt\fR(4fe) file to
standard output.  If no \fIfilename\fR is given then the standard input is
used.  Any syntactic or semantic errors are reported to standard error.  A
complete description of the mathematics can be found in the user's guide.
.SH OPTIONS
\fICorduroy\fR accepts the following options:
.TP 5
.B \-help
Write a message describing valid command line options to standard error
and exit.
.TP
.B \-debug
Write the processed file to standard error in the form of a valid input
file.
.TP
.B \-nocpp
Do not use a preprocessor on the input file.
.TP
.BI \-cpp " filename"
Use \fIfilename\fR as a preprocessor on the input file.  The default
preprocessor is "/lib/cpp".  Any preprocssor which understands the -D, -U,
and -I options can be used as these options are passed to the preprocessor.
.SH AUTHOR
\fICorduroy\fR was developed by Jason I. Gobat (jgobat@mit.edu) and Darren
C. Atkinson (atkinson@ucsd.edu).
.SH SEE ALSO
felt(1fe), velvet(1fe), corduroy(4fe), felt(4fe), burlap(1fe).


CHANGELOG
** Changes to FElt from 3.02 to 3.05 (first version released on SourceForge)

The following changes were made to the FElt syntax:
   1. changed the sense of coordinates for axisymmetric elements to
      make them easier to use with mesh generation.  x is still radial
      but now y is vertical/axial.  Previously z was vertical.  

The following bugs/things were fixed/changed:
   1. some small bugs in axisymmetric, ctg, htk, and iso_2d elements
      that I haven't been really good about keeping tack of.  Try
      the new version and see if your answers look better -- if they
      look worse then let me know ...
   2. you can now use the -details switch to get the coordinates of
      the points where stresses are calculated in the text display of
      results.
   3. added radial and axial directions to the load dialog in velvet
      and to the file writing routines.  They'd gotten left out when
      axisymmetric elements were originally implemented.
   4. finally resolved the colormap issue in velvet so you can 
      use color stress and displacement plots on > 8-bit displays.
      Take a look at src/Velvet/XCC.README to see how you too can
      accomplish this if you're X11 colormap impaired like us.
   5. dropped Geompk in favor of Triangle by J.R. Shewchuk.  Several
      reasons for the change: Geompk was always a little finicky,
      mostly because of our interface probably, but it was sometimes
      difficult to coax a mesh out of Geompk.  The interface to
      Triangle is very simple, it is coded in C which makes compilation
      and configuration on other architectures simple, and it seems
      to be pretty easy to get a decent mesh out of it.  Accordingly,
      the main input parameters have changed to target and alpha.
      target is the approximate number of desired elements in the mesh.
      alpha is used to generate a maximum bound on the area of a generated
      element: Amax of an elt = alpha*Atotal/target.  Triangle appears to
      have a lot of other options for mesh refinement, angle constraints
      etc., but none of these are made use of at this point.  Most of the
      hooks for Geompk are still available if someone really wants to
      switch back.  The call to GenerateTriMesh did not change for 
      applications, but the applications did change to account for
      the loss of Geompk parameters (min, max, angspc, angtol, etc.)
      and the addition of target and alpha.
     
** Changes to FElt from 3.00 or 3.01 to 3.02 

The following changes were made to the FElt syntax:
   1. hooks for a gravity vector in the analysis section.  Doesn't
      do anything yet, but the syntax is there ...
   2. "static-substitution" and "static-incremental" are now valid 
      analysis types
   3. a "load cases" section and accompanying syntax for named
      load cases
   4. syntax for new analysis parameters, "tolerance", "iterations",
      "relaxation", "load-steps".
   5. input-node= and input-dof= parameters in the analysis 
      parameters to allow for parametric studies in which you
      get a table (or plot) of output that explores displacements
      as a function of force level at a single input point
   6. axial and radial are now valid directions for distributed loads

The following bugs/things were fixed/changed:
   1. the transform matrix for beam3d elements was incorrect if elements
      had nodes which were out of the x-y plane.
   2. several problems with the new Makefile scheme were corrected.
   3. velvet was probably getting the aspect ratios wrong in drawing 
      3d structures due to an error in the min/max checking.
   4. the spectrum components of force structures were not being
      initialized to either zero or NULL
   5. save's in velvet were actually doing dumps (save all's), minor, 
      but potentially annoying.
   6. there was still a problem with the lumped mass matrices for 
      htk elements.
   7. the new PrintGlobalMatrices routine took a file pointer but
      didn't actually use it; things were just going to stdout.
   8. internal error in burlap's interface wit the matrix lib
      was causing some problems with inversion and overwriting.
   9. there were problems with iso_2d elements (or any element with
      optional nodes) as the renumbering routines were just blindly
      looping through element -> definition -> numnodes rather than
      checking to make sure that all of the nodes were in fact valid.
  10. on a similar note, the checks in velvet for out-of-plane nodes
      were going over element -> definition -> numnodes rather than
      element -> definition -> shapenodes (which all have to be valid
      in the first place).
  11. there was some sort of problem with the pre-processor and
      corduroy (don't ask me how these things happen).  I think it's
      fixed now, but then again, I don't really know what was wrong.
  12. renamed AssembleStiffness and AssembleDynamic to ConstructStiffness
      and ConstructDynamic to reflect that they both create and assemble
      the matrices (this then parallels ConstructForceVector).  Renamed
      ConstructTransientForce to AssembleTransientForce to reflect 
      the fact it only fills the thing rather than creating it then
      filling it.
  13. the cpp command line wasn't being built properly so you couldn't
      actually use -I, -D and -U on the felt (and corduroy) command line
  14. memory initialization problem in Drawing.c
  15. 2d and 3d Athena widget specific defaults were moved into
      velvet.c and are now applied automatically at runtime as
      appropriate.
  16. the filename printing and scrolling in velvet's file dialog
      was improved.
  17. changed the analysis.nodes array into an array of actual node
      pointers rather than node numbers.  In some ways this adds
      a layer of complexity because whenever we input analysis nodes
      we have to resolve their numbers by doing tree lookups in the

src/Corduroy/corduroy.c
/*
    This file is part of the FElt finite element analysis package.
    Copyright (C) 1993-1997 Jason I. Gobat and Darren C. Atkinson

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

/************************************************************************
 * File:	corduroy.c						*
 *									*
 * Description:	This file contains the code the driver for the corduroy	*
 *		application.						*
 ************************************************************************/

# include <stdio.h>
# include <ctype.h>
# include <string.h>
# include "allocate.h"
# include "mesh.h"
# include "generator.h"
# include "error.h"
# include "definition.h"


# define streq(a,b)	!strcmp(a,b)


# ifndef DOS
static char *usage = "\
usage: corduroy [options] [filename]\n\
       -debug           write debugging output\n\
       -nocpp           do not use a preprocessor\n\
       -cpp filename    preprocessor to use\n\
       -Dname[=value]   define a macro\n\
       -Uname           undefine a macro\n\
       -Idirectory      specify include directory\n\
";
# else
static char *usage = "\
usage: corduroy [options] [filename]\n\
       -debug           write debugging output\n\
";
# endif


/************************************************************************
 * Function:	Quote							*
 *									*
 * Description:	Quotes a string if necessary.				*
 ************************************************************************/

static char *Quote (s)
    char *s;
{
    char	c;
    char       *ptr;
    static char buffer [256];


    for (ptr = s; (c = *ptr); ptr ++)
	if (!isalpha (c) && c != '_' && (isdigit (c) ? ptr == s : 1)) {
	    sprintf (buffer, "\"%s\"", s);
	    return buffer;
	}

    return s;
}


/************************************************************************
 ************************************************************************/

static void AddNewArrays (node, element, numnodes, numelts,
                          all_nodes, all_elements, nn, ne)
   Node		*node;
   Element	*element;
   unsigned	numnodes, numelts;
   Node		**all_nodes;
   Element	**all_elements;
   unsigned	*nn;
   unsigned	*ne;
{
   unsigned	i;

   ZeroOffset (*all_nodes);
   *all_nodes = Reallocate (*all_nodes, Node, *nn + numnodes);
   UnitOffset (*all_nodes);

   ZeroOffset (*all_elements);
   *all_elements = Reallocate (*all_elements, Element, *ne + numelts);
   UnitOffset (*all_elements);

   for (i = 1 ; i <= numnodes ; i++) 
      (*all_nodes) [i + *nn] = node [i];

   for (i = 1 ; i <= numelts ; i++) 
      (*all_elements) [i + *ne] = element [i];

   *nn += numnodes;
   *ne += numelts;
}



/************************************************************************
 * Function:	main							*
 *									*
 * Description:	Main the the driver function for the corduroy		*
 *		application.						*
 ************************************************************************/

int main (argc, argv)
   int		argc;
   char		*argv [ ];
{
   Definition	definition;
   Element	*element;
   Node		*node;
   unsigned	start_node;
   unsigned	start_element;
   unsigned	numnodes;
   unsigned	numelts;
   Line		*line;
   Grid		*grid;
   Grid		*quadgrid;
   Grid		*brickgrid;
   TriMesh	*trimesh;
   unsigned	i,j;
   unsigned	nn, ne;
   Node		*all_nodes;
   Element	*all_elements;
   int		 debug = 0;
   unsigned	 code;


	/*
	 * check usage and read the input file
	 */

# ifndef DOS
   if (CorduroyCppOptions (&argc, argv)) {
	fputs (usage, stderr);
	exit (1);
   }
# endif

   j = 1;
   for (i = 1; i < argc; i ++) {
	if (streq (argv [i], "-help")) {
	    fputs (usage, stderr);
	    exit (0);
	} else if (streq (argv [i], "-debug"))
	    debug = 1;
	else
	    argv [j ++] = argv [i];
   }

   argv [argc = j] = NULL;

   if (argc > 2) {
	fputs (usage, stderr);
	exit (1);
   }

   add_all_definitions ( );
   if (ReadCorduroyFile (argc == 2 ? argv [1] : "-"))
	exit (1);
   
   if (debug) {
	WriteCorduroyFile ("-");
        exit (1);
   }

	/*
	 * initialize some arrays and counters
	 */

   line = generator.lines;
   grid = generator.grids;
   quadgrid = generator.quadgrids;
   brickgrid = generator.brickgrids;
   trimesh = generator.trimeshes;

   nn = 0;
   ne = 0;
   start_node = generator.start_node - 1;
   start_element = generator.start_element - 1;
   
   all_nodes = NULL;
   all_elements = NULL;

	/*
	 * generate along lines
	 */

   for (i = 0 ; i < generator.num_lines ; i++) {
      GenerateLine (line[i], &element, &node, 
                    &numelts, &numnodes, start_node, start_element);

      start_node += numnodes;
      start_element += numelts;

      AddNewArrays (node, element, numnodes, numelts, 
                    &all_nodes, &all_elements, &nn, &ne);

      ZeroOffset (node); Deallocate (node);
      ZeroOffset (element); Deallocate (element);
   }

	/*
	 * generate grids of line elements
	 */

   for (i = 0 ; i < generator.num_grids ; i++) {
      GenerateGrid (grid[i], &element, &node, 
                    &numelts, &numnodes, start_node, start_element);

      AddNewArrays (node, element, numnodes, numelts, 
                    &all_nodes, &all_elements, &nn, &ne);

      start_node += numnodes;
      start_element += numelts;

      ZeroOffset (node); Deallocate (node);
      ZeroOffset (element); Deallocate (element);
   }

	/*
	 * generate grids of quadrilateral elements
	 */

   for (i = 0 ; i < generator.num_quadgrids ; i++) {
      GenerateQuadGrid (quadgrid[i], &element, &node, 
                        &numelts, &numnodes, start_node, start_element);

      AddNewArrays (node, element, numnodes, numelts, 
                    &all_nodes, &all_elements, &nn, &ne);

      start_node += numnodes;
      start_element += numelts;

      ZeroOffset (node); Deallocate (node);
      ZeroOffset (element); Deallocate (element);
   }

	/*
	 * generate grids of brick elements
	 */

   for (i = 0 ; i < generator.num_brickgrids ; i++) {
      GenerateBrickGrid (brickgrid[i], &element, &node, 
                         &numelts, &numnodes, start_node, start_element);

      AddNewArrays (node, element, numnodes, numelts, 
                    &all_nodes, &all_elements, &nn, &ne);

      start_node += numnodes;
      start_element += numelts;

      ZeroOffset (node); Deallocate (node);
      ZeroOffset (element); Deallocate (element);
   }

	/*
	 * generate meshes of triangular meshes
	 */

   for (i = 0 ; i < generator.num_trimeshes ; i++) {
      code = GenerateTriMesh (trimesh[i], &element, &node, 
                              &numelts, &numnodes, start_node, start_element);
      if (code)
         continue;

      AddNewArrays (node, element, numnodes, numelts, 
                    &all_nodes, &all_elements, &nn, &ne);

      start_node += numnodes;
      start_element += numelts;

      ZeroOffset (node); Deallocate (node);
      ZeroOffset (element); Deallocate (element);
   }

	/*
	 * coalesce the nodes
	 */

   if (nn == 0 || ne == 0)
	exit (1);

   all_nodes = CoalesceNodes (all_nodes, all_elements, &nn, ne);

	/*	
	 * write everything out
	 */

   printf ("problem description\n");
   printf ("nodes=%d elements=%d\n\n", nn, ne);

   printf ("nodes\n");
   if (nn)
	if (generator.constraint)
	    printf ("%d  x=%g   y=%g   z=%g constraint=%s\n",
	      all_nodes [1] -> number, all_nodes [1] -> x, all_nodes [1] -> y,
	      all_nodes [1] -> z, Quote (generator.constraint));
	else
	    printf ("%d  x=%g   y=%g   z=%g\n", all_nodes [1] -> number, 
              all_nodes [1] -> x, all_nodes [1] -> y, all_nodes [1] -> z);

   for (i = 2 ; i <= nn ; i++) 
      printf ("%d  x=%g   y=%g   z=%g\n", all_nodes [i] -> number, 
              all_nodes [i] -> x, all_nodes [i] -> y, all_nodes [i] -> z);

   definition = all_elements [1] -> definition;
   printf ("\n%s elements\n", definition -> name);

   for (i = 1 ; i <= ne ; i++) {
      if (all_elements [i] -> definition != definition) {
         definition = all_elements [i] -> definition;
         printf ("\n%s elements\n", definition -> name);
      }

      printf ("%d   nodes=[%d,", all_elements [i] -> number, 
              all_elements [i] -> node[1] -> number);

      for (j = 2 ; j < definition -> numnodes ; j++) 
         printf ("%d,", all_elements [i] -> node[j] -> number);

      printf ("%d]",all_elements[i] -> node[definition->numnodes] -> number);
      if (i == 1 && generator.material)
	printf (" material=%s", Quote (generator.material));
      printf ("\n");
   }
      
   printf ("\nend\n");
   return 0;
}
